### PR TITLE
Reduce gap between snippets and generic views/templates

### LINKED ISSUE #11393: No permission check for "why not add one?" link in snippets index view
### Issue Summary
Two templates in the snippets app do not check permissions before rendering a link to the create view for for the snippet model.

In the index results when there are no entries:
https://github.com/wagtail/wagtail/blob/8a6ab175fc1ee2eab53f7223a5bb24e97a2a278f/wagtail/snippets/templates/wagtailsnippets/snippets/index_results.html#L17-L21

In the chooser:
https://github.com/wagtail/wagtail/blob/8a6ab175fc1ee2eab53f7223a5bb24e97a2a278f/wagtail/snippets/templates/wagtailsnippets/chooser/results.html#L8-L11

I'm not sure what the policy is in Wagtail but I think it is counter-intuitive to render a link to a create form when the user does not have privileges to do so.

### PR DESCRIPTION
**Pull request by laymonage**
Fixes #11393 ~~(by removing the link altogether...)~~

Towards #11366 (not all of the views in the issue are fixed, to keep the PR scope minimal).

- Use `BaseListingView` (and its template) instead of `generic.IndexView` when possible:
  - `ModelIndexView`
  - Usage view
  - History (w. results) view
- Move remaining "optional features" attributes to the optional features mixin
- Add status side panel to the generic Create view. This isn't very useful at the moment, but it helps for consistency with edit view and allows the snippet CreateView to reuse more code. This will be useful when we add support for switching locale in the generic CreateView, and if we want to add "model description" similar to `page_description` in the future.
- Add default `error_message` and `success_message` to the generic `Create`/`Edit`/`Delete` views.
- Reduce overrides in snippets CRUD view and templates that are no longer necessary.

I'll have a follow-up PR (#11445) to add proper support for locale switching in the generic CreateView. That should help translation libraries when working with `ModelViewSet`s.

Once both PRs are in, the remaining overrides in the snippets views are:
- To add support for snippets hooks (backwards-compatibility)
- To add bulk actions support (to be addressed in #11371)
- To add base queryset customisation support (to be addressed in #10746)
- To add footer action menu
- To add optional features side panels (i.e. the `SnippetStatusSidePanel` and `PreviewSidePanel`, to be addressed in #11358)
- A few minor bits remaining in #11366

Once all of the above are addressed:
- Generic features should be available in `ModelViewSet` and their generic views by default.
- View logic for features that are based on whether the model has certain mixins (e.g. `RevisionMixin`) are put in the `FooOptionalFeaturesMixin` in `wagtail.admin.views.generic.mixins`.
  - These mixins refer to views that are available generically but not enabled by default in `ModelViewSet` (but are enabled in `SnippetViewSet`), e.g. revisions compare view
- Snippets-specific legacy things e.g. hooks and footer actions are kept in the snippets views, and they should be the only overrides left

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

I have a bakerydemo branch that uses `ModelViewSet` to help me test this: https://github.com/laymonage/bakerydemo/tree/country-modelviewset, but I also test with other ways e.g. changing the `FooterText` to use `ModelViewSet` (just replace `SnippetViewSet` with it) and running the testapp locally.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

File: client/scss/components/_footer.scss
Comment (laymonage) [inline client/scss/components/_footer.scss:172]: The footer assumes it's not in an element with `nice-padding` applied, as it has its own margin with nice-padding's values (see lines above this). With the way the generic views are set up, allow unsetting the margin so the footer actions do not have extra offset.

This will probably be made redundant with Thibaud's work on #9030 when that comes along.

File: wagtail/snippets/templates/wagtailsnippets/snippets/create.html
Comment (laymonage) [inline wagtail/snippets/templates/wagtailsnippets/snippets/create.html:24]: Before merging, might want to wait until #9030 is implemented in case that work becomes hard to merge due to this change

File: wagtail/snippets/templates/wagtailsnippets/snippets/create.html
Comment (laymonage) [inline wagtail/snippets/templates/wagtailsnippets/snippets/create.html:8]: #11358

File: wagtail/snippets/templates/wagtailsnippets/snippets/index_results.html
Comment (laymonage) [inline wagtail/snippets/templates/wagtailsnippets/snippets/index_results.html:10]: I opted to remove this, but we can also move this to the generic templates if desired

Comment (gasman) [reply wagtail/snippets/templates/wagtailsnippets/snippets/index_results.html:10]: I'd prefer to see this reinstated in the generic templates - or at least we should have better justification for removing it than "it saves us having to fix #11393" :-). I have no strong feelings about the message personally, but I think it's better that we don't casually remove things that people might be accustomed to, or have built workflows around.

Besides that, these changes look good to me!

File: client/scss/components/_footer.scss
Comment (laymonage) [inline client/scss/components/_footer.scss:25]: Found a nicer way to fix this without having to add `.footer--no-margin`

File: wagtail/admin/locale/ca/LC_MESSAGES/django.po
Comment (laymonage) [inline wagtail/admin/locale/ca/LC_MESSAGES/django.po:698]: Not sure how we feel about manually editing translation files, but I think we should try not to add extra work for translators. These are all in a separate commit d2da6b49be5439331d331e3791d3fc98b66f4f22 though, so if it's undesired I can drop it.

File: wagtail/admin/templates/wagtailadmin/generic/index_results.html
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/generic/index_results.html:42]: Not sure if this is enough context for the translators though. I guess I can also make this an if/else with two translatable blocks, e.g. having
- `"There are no {{ model_name }} to display.` as one block
- `"There are no {{ model_name }} to display. Why not <a href="{{ add_url }}">add one</a>?"` as the other.

Or maybe rename the variable to `add_model_url`? Not sure if that helps much

File: wagtail/admin/templates/wagtailadmin/generic/history_results.html
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/generic/history_results.html:1]: Unfortunately can't change this to `listing_results.html`, since that template doesn't support the active filters yet. That functionality came for free when it was implemented in the index view, and changing this to `listing_results.html` means we'll lose that.

In addition, the `listing_results.html` template doesn't have the "There are x results" message, so either we need to add that there, or update the tests to not expect it. In any case, that will probably be removed in the future, in favour of showing the number of results in the heading of the first table column like we do for pages.

I opted to leave this in for now.

File: wagtail/admin/views/generic/history.py
Comment (laymonage) [inline wagtail/admin/views/generic/history.py:137]: No longer needed after 5ede8a1 which uses the new slim header buttons.

File: wagtail/admin/views/generic/history.py
Comment (laymonage) [inline wagtail/admin/views/generic/history.py:134]: The results template still uses `index_results.html` but now without the `no_results_message` block override, so we need this in order to show `"There are no {{ model_name }} to display."`

We can also use the queryset's `.model._meta`, but that means it'll read as `"There are no model log entries to display"` which seems too specific for end-users, especially for models with custom log entry models. With the `BaseLogEntry`, it just shows as `"There are no log entries to display."`, which is quite similar to the previous `no_results_message` block override.

File: wagtail/admin/views/generic/models.py
Comment (laymonage) [inline wagtail/admin/views/generic/models.py:341]: Fixes #11393

**Comment by laymonage** [comment]
@gasman I've reinstated the "Why not add one?" link in 30ab8825b5b8fbe13e41bf847d42589fac1a069e and added test in ffb349ead28ec898d86e8ba8e46a1c4e4b9a0cbf to ensure the issue is fixed, although I had to make a few minor tweaks in order to do that (and to resolve conflicts after rebasing). The last 5 commits are new, while e60ce790d751cf886df6b652bc057da72ba35279 had to be amended after the active filters were implemented

File: wagtail/admin/locale/ca/LC_MESSAGES/django.po
Comment (gasman) [reply wagtail/admin/locale/ca/LC_MESSAGES/django.po:698]: Happy to keep this in, but we'll need to think about how to push this change to Transifex. I _think_ fetching translations, then rebasing/merging this PR on top, then `tx push` should do it...

File: wagtail/admin/templates/wagtailadmin/generic/index_results.html
Comment (gasman) [reply wagtail/admin/templates/wagtailadmin/generic/index_results.html:42]: I think I'd go with the "two translatable blocks" solution. I notice that there are various other result templates such as [wagtailredirects/results.html](https://github.com/wagtail/wagtail/blob/main/wagtail/contrib/redirects/templates/wagtailredirects/results.html) that have the same text without the `model_name` variable (and probably need the same permission check, but at least we can consider that out of scope for #11393...) - in those cases it's definitely better to have both sentences in the same string, to allow the "add one" text to be adjusted to fit the model name (e.g. matching gender). It makes sense for the snippet template to follow the same pattern, even if that particular advantage isn't present here.

File: wagtail/admin/locale/ca/LC_MESSAGES/django.po
Comment (laymonage) [reply wagtail/admin/locale/ca/LC_MESSAGES/django.po:698]: Okay, I'll do that when I merge (or leave it to you if you'd like to merge)

File: wagtail/admin/templates/wagtailadmin/generic/index_results.html
Comment (laymonage) [reply wagtail/admin/templates/wagtailadmin/generic/index_results.html]: OK, added in 9da5434. Curious what we'd prefer when we eventually refactor those views though (i.e. whether to keep the original block overrides or reuse the generic one to unify the code), but we'll cross that bridge when we come to it.

**Review by gasman — APPROVED** [review]
Thanks for the updates - all looks good to me now!

File: wagtail/admin/locale/es/LC_MESSAGES/django.po
Comment (laymonage) [inline wagtail/admin/locale/es/LC_MESSAGES/django.po:805]: Not sure why this and a few other files weren't replaced when bc6ef435225c5f3a37a5e9ccdcaa59a2cfcb6e2e was done

**Review by laymonage — COMMENTED** [review]
Have fetched new translations on top of latest main, rebased this PR, and done `tx push`. Should be all good now, merging.

**Pull request merged**
